From: Martin Pitt <martin.pitt@ubuntu.com>
Date: Wed, 11 Mar 2015 10:37:47 +0100
Subject: systemctl: Don't forward telinit u to upstart

When dist-upgrading and /sbin/init changes from upstart to systemd, upstart's
Restart() exec's argv[0] which is  still /sbin/init, and thus it execs systemd.
This is triggered by glibc upgrades or other things which call "telinit u" in
their postinst scripts.

This should eventually be fixed in upstart itself, but as we can't
retroactively fix it in previous releases we upgrade from, we simply ignore
the request.

https://launchpad.net/bugs/1430479
---
 src/systemctl/systemctl.c | 7 +++++++
 1 file changed, 7 insertions(+)

--- a/src/systemctl/systemctl.c
+++ b/src/systemctl/systemctl.c
@@ -8334,6 +8334,13 @@
                                 /* Hmm, so some other init system is running, we need to forward this request to
                                  * it. For now we simply guess that it is Upstart. */
 
+                                /* work around upstart exec'ing systemd when /sbin/init
+                                 * changes (https://launchpad.net/bugs/1430479) */
+                                if (argv[1] != NULL && streq(argv[1], "u")) {
+                                    log_warning("Ignoring telinit u request, systemd is not running");
+                                    return -ENOTSUP;
+                                }
+
                                 (void) rlimit_nofile_safe();
                                 execv(TELINIT, argv);
 
